### ISSUE TITLE
super slow: cheerio struggling with some selectors

### ISSUE THREAD

**Issue by dbuezas** [cheeriojs/cheerio#1118]
Cheerio gets super slow (minutes per query) with selectors of this form: 
`'.row div div div div div div div span'`
while Sizzle (with and without document.querySelectorAll available) stay fast

weirdly enough, this is instantaneous:
`'* div div div div div div div span'`

I'm using this page as an example: https://www.bestbuy.com/site/lenovo-ideapad-11-6-laptop-intel-celeron-2gb-memory-32gb-emmc-flash-memory-mineral-gray/6061800.p?skuId=6061800
But the same happens in most html pages.

I'm guessing this may be a drawback from the way css-select processes the query, because I see that the same checks get done over the same elements thousands (maybe millions) of times (e.g.  checking if a particular element has the `row` class). 

If the query is divided like this `$('.row').find('div').......find('span') ` the problem disappears.

Any ideas?

**Comment by dbuezas** [comment]
Oh, I think I see the problem in css-select's way of resolving the queries. This whole stack blows my mind by the way (props to the devs!).
I think what is happening is that the 'descendant' token (empty space in selector) makes the rest of the query (from right to left) to be tested at every level.

So if we have a tree of height 10 and this 3 parts selector `.row div div` the search will start from a leaf and at every step going up it can 'accept' an element and move on to the next token (the next selector part to the left) or skip that element and try the whole thing on the parent. 
This means there are two options at each level (accept element and move on to the next selector part to the left OR jump to the parent element), and for all non descendants of `.row`, the two options have to be explored at each level since that part of the query will fail at the end in every case. 

TL;DR
If I didn't get everything wrong, the time complexity may be O(n) on the number of elements, but if one considers the selector then it may transform into something more like this `O(n*2^(h*d))`, with `h=height of the tree` and `d=amount of descendant tokens in selector`. And that mostly because of the lack of caching of the results.

**Comment by dbuezas** [comment]
Is confirmed, I found a solution and will create a PR to the css-select repo.

**Comment by fb55** [comment]
Moving to css-select. Thanks for investigating!
